### PR TITLE
[Merged by Bors] - feat(Algebra/Irreducible): prove irreducibility preserved by `MulEquiv`

### PR DESCRIPTION
**Pull request by jjdishere**
Add two lemmas about Irreducibility preserved by `MulEquiv`.

---

- [x] depends on: #17403 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Associated/Basic.lean [hunk 1]
```diff
@@ -290,6 +290,21 @@ theorem irreducible_mul_iff {a b : α} :
...
     · rwa [irreducible_isUnit_mul ha]
 
+variable [Monoid β] {F : Type*} [EquivLike F α β] [MulEquivClass F α β] (f : F)
+
+open MulEquiv
+
+theorem Irreducible.map {x : α} (h : Irreducible x) : Irreducible (f x) :=
+  let f := MulEquivClass.toMulEquiv f
+  ⟨fun g ↦ h.1 (symm_apply_apply f x ▸ g.map f.symm), fun a b g ↦
+    .elim (h.2 _ _ (symm_apply_apply f x ▸ map_mul f.symm a b ▸ congrArg f.symm g))
```
Comment (eric-wieser) [inline Mathlib/Algebra/Associated/Basic.lean:300]: Let's wait for #17089 here.

**Comment by jjdishere** [comment]
This PR is ready for review again.

**Comment by jjdishere** [comment]
In fact, after #17089 is merged still we cannot generalize the `Irreducible.map` to a more general setup. Mathematically, one really needs a bijection to prove this theorem.

**Comment by riccardobrasca** [comment]
Surjective + local hom is not enough, right?

**Comment by riccardobrasca** [comment]
> > Surjective + local hom is not enough, right?
> 
> Not enough. In order to show `f (x) = a * b` implies `a` or `b` is a unit, one needs to find preimage `a'` and `b'` of `a` `b` such that `a' * b' = x` and use the fact that `x` is irreducible. However, surjectivity cannot guarantee this. It only guarantees `a' * b' = x * (something in the kernel)`.

Yeah, I see it doesn't work immediately, but it would be nice to have an easy counterexample for the docstring. Anyway it is not very important. Thanks!

bors merge

**Comment by jjdishere** [comment]
> Yeah, I see it doesn't work immediately, but it would be nice to have an easy counterexample for the docstring. Anyway it is not very important. Thanks!

I found a counterexample and added it to the docstring.

**Comment by riccardobrasca** [comment]
> > Yeah, I see it doesn't work immediately, but it would be nice to have an easy counterexample for the docstring. Anyway it is not very important. Thanks!
> 
> I found a counterexample and added it to the docstring.

Nice!

bors merge

**Pull request closed without merging**